Add debug feature flag

FlyingRatBull opened the pull request:
* Added feature flag for debug renders (enabled by default)
* Added field in PluginState to toggle debugging during runtime

aevyrie commented on src/lib.rs:
```diff
@@ -52,18 +68,45 @@ impl<T: 'static + Send + Sync> Default for DefaultRaycastingPlugin<T> {
+
+impl<T> PluginState<T> {
+    #[cfg(feature = "debug")]
+    pub fn new(enabled: bool, debug: bool) -> Self {
```
I'd like to replace the booleans here with some enums to make this more explicit. Something like 

```rust
pub fn new(state: State, debug: Debug) -> Self {
```

where

```rust
pub enum State {
    Enabled,
    Disabled,
}

pub enum Debug {
    Enabled,
    Disabled,
}
```

That way, use of `new()` looks like this in practice:

```rust
commands.insert_resource(PluginState::<MyRaycastSet>::new(State::Enabled, Debug::Disabled));
```

FlyingRatBull replied:
Good idea, I will implement it.
A thing to be aware of though is, that moving the enabled state from `bool` to `enum` is a braking change. However I already bumped the version in this PR to `0.3.0` so this would be reflected.

aevyrie replied:
Yep! I think the version bump is sufficient. Thanks!

aevyrie reviewed (changes requested):
Hey @FlyingRatBull, this is a great addition, thank you. I only have one requested change. Other than that, it looks like there are a few tiny things to fix in CI.

FlyingRatBull commented:
Ok... after some refactorings I pretty much reimplemented the whole Debug Cursor functionality. It is now much more streamlined and has better performance.
Your suggestions regarding enums instead of bools are incorparated.

**Please be aware that I also fixed a memory leak. The current version on crates.io should be yanked and `bevy_mod_picking` (as well as other projects depending on this crate) should update to the new version!**

Modifications:
* Added `debug` feature (previous commit)
* Fixd memory leak in `update_debug_cursor` which created new meshes each frame
* Each `RayCastSource` stores one potential debug cursor (feature-gated)
* Debug cursor get added/removed when their `RayCastSource` counterparts get added/removed
* Debug cursor get added/removed when the plugin state (enabled/debug enabled) changes
* New debug states `None` and `Cursor` allow for future extensions such as lines from the source to the target
* Optimized `update_debug_cursor` to only handle the one cursor associated with the `RayCastSource`
* Extended `mouse_picking` example:
  * Added another source which can be moved with the arrow keys
  * Spacebar adds/removes the Keyboard-controlled source
  * Left click enables/disables ray casting
  * Right click enables/disables debug cursors

aevyrie commented:
Hey @FlyingRatBull, sorry to keep you waiting. Can you explain the memory leak you resolved? AFAIK, meshes are ref counted, so continuously adding a component with a mesh every frame will:

1. Overwrite the existing component
2. The old mesh will be dropped now that its handle is no longer referenced

If those assumptions are _not_ true, this is something that also needs to be improved upstream in Bevy, whether it's docs or some other change.

FlyingRatBull commented:
Hey @aevyrie, you don't happen to know the part in the documentation about `Handle` lifetimes? I found the relevant part in the [Unofficial Bevy Cheat Book](https://bevy-cheatbook.github.io/features/assets.html#handles-and-asset-lifetime-garbage-collection) but had no look with the official documentation.

The code which produces the leak is:

```rust
let debug_material = &materials.add(StandardMaterial {
        base_color: Color::rgb(0.0, 1.0, 0.0),
        unlit: true,
        ..Default::default()
});
```

I also debugged it further and reached memory usage in the Gigabyte range, so it doesn't seem to be an issue with latent garbage collection.

If you'd be so kind as to point me to the official docs, or I find the relevant part in the meantime I will of course create an upstream issue for either the code or the docs to be updated.

aevyrie commented:
> If you'd be so kind as to point me to the official docs, or I find the relevant part in the meantime I will of course create an upstream issue for either the code or the docs to be updated.

I think this is just information I picked up from the discord server over time, but my understanding may be flawed. We could probably create a minimal reproduction by creating a new material every time a system is run in a simple demo like [this](https://github.com/bevyengine/bevy/blob/main/examples/3d/3d_scene.rs)

aevyrie commented on src/lib.rs:
```diff
@@ -52,18 +84,61 @@ impl<T: 'static + Send + Sync> Default for DefaultRaycastingPlugin<T> {
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum Enabled {
```
I'd prefer to name this something like `ActiveState` or something. Then the `PluginState` would be something like
```rust
if plugin.active == ActiveState::Enabled {
// ...
```

FlyingRatBull replied:
Done

aevyrie commented on src/lib.rs:
```diff
@@ -52,18 +84,61 @@ impl<T: 'static + Send + Sync> Default for DefaultRaycastingPlugin<T> {
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum Debug {
```
To mirror the above comment, it might make sense to name this `DebugState` to make them symmetric.

FlyingRatBull replied:
Done

aevyrie commented on examples/mouse_picking.rs:
```diff
@@ -1,6 +1,10 @@
 use bevy::prelude::*;
```
This is a really cool demo, but I'd like to keep the `mouse_picking` example as minimal as possible to make it clear what the absolute minimum is required to get it working.

However, this updated example shows how to accomplish something a bit more complicated and representative. Let's keep the old version of `mouse_picking` as-is, and rename what you've created here to `multiple_sources`.

FlyingRatBull replied:
Done

aevyrie commented on src/bounding.rs:
```diff
@@ -37,9 +38,10 @@ pub fn update_bound_sphere<T: 'static + Send + Sync>(
     >,
 ) {
-    if !state.enabled {
+    if state.enabled == crate::Enabled::Disabled {
```
See comments in lib.rs, I feel like the enum being named `Enabled` is a bit confusing in situations like this. As mentioned in that comment, the suggested change would change this into:
``` rust
if state.active == ActiveState::Disabled {
// ...
```

FlyingRatBull replied:
Done

aevyrie commented on src/debug.rs:
```diff
@@ -1,129 +1,248 @@
+    commands: &mut Commands,
+    source_entity: Entity,
+    mut source: &mut RayCastSource<T>,
+    sources_to_cursors: &mut ResMut<SourceToCursorMap<T>>,
```
Do we need to maintain a separate resource that tracks entity pairs? Wouldn't it make more sense to add a component onto the `RayCastSource` itself? In that case there is no need to maintain a map, instead the `DebugCursor` just looks at its parent (a `RayCastSource`) to know where it needs to go.

While I don't think the current implementation in `master` is ideal, especially because it just sticks the debug cursor mesh right in the Camera entity, it's pretty simple to reason about, which I really value for this library.

I could be convinced your implementation is the right choice, I'm just wary of adding complexity unless we really need it.

FlyingRatBull replied:
Sorry for the late response, I've been on vacation.

While I agree with you on the account of simplicity I cannot find a simple way to query the parent/child relationship with a specific component. That means that I would have to iterate over all child entities of the RayCastSource (which could be the camera/player with many children) and manually check, if this entity has the DebugCursor component.
As the debugging functionality is likely to be used in debugging use cases, the performance of the already slow debug build (in general) would get even worse.

IMO the gained performance outweighs the complexity in this case, but if you think the impact could be tollerated in these cases, I can modify it accordingly of course.

aevyrie replied:
>Sorry for the late response, I've been on vacation.

No worries 🙂.

> I cannot find a simple way to query the parent/child relationship with a specific component. [...]

I think you can do this a few ways, one way I might approach it:

```rust
// ...
parent_query: Query<RayCastSource>,
child_query: Query<DebugCursor>
// ...

for parent in parent_query.iter() {
    // Use the debug entity stored in the parent to get that DebugCursor entity
    child_query.get(parent.debug_entity)
}
```

aevyrie commented:
I took some time to review this in a bit more detail, and I've come to the conclusion that this needs to be split into a few PRs. This is (relative to the repo) a very large PR, and I disagree with some of the premises you've presented.

With that said, there is a lot here that I really like! I'd love to get the following things merged, but it will be much easier if you split them into separate PRs that we can review and merge independently:

1. Gating debug functionality behind a feature flag.
2. Improvements to the granularity and explicitness state with `ActiveState` and `DebugState`.
3. The added `multiple_sources` example.

Things that need more discussion:

1. Claims of performance improvement/regression. Without some sort of benchmark (even an imperfect one!), I can't trust that changes to debug cursor placement have any measurable effect. Plus, the hot path in this codebase is in the ray casting and culling - I'm not super worried about the debug cursor. I get ~500fps using integrated graphics in the `minimal` example with vsync off.
2. Memory leaks. I did some testing on the master branch, and verified that the `Mesh` resource never grows as debug meshes are created naively. With the `minimal` example, the number of `Mesh` resources is constant at 4, never increasing. There may be value in storing a handle to the mesh instead of generating it every frame, but we should figure out what the right solution is before jumping to code.
3. Anything that adds complexity. This plugin is depended on by other plugins and applications. I really want to keep it as simple and flexible as possible. I have a few plugins that track Bevy/main, the less code I have to maintain, the better. 🙂 

In the future, opening an issue and discussing first, especially for large changes, will make things easier for both of us. 🙂

aevyrie commented:
Closing due to age
